### PR TITLE
Implement new TxDatastore and Txn interfaces

### PR DESCRIPTION
**Pull request by bigs**
adds full on transaction support for the badger datastore. related to ipfs/go-datastore#91, though not dependent on it. also updates all vanilla Datastore methods to leverage the transaction wrapper.

WIP for now because it doesn't include transaction specific tests, though all existing tests (which depend on them implicitly) are passing.

### CODE REVIEW

**Comment by bigs** [comment]
added TTL!

**Comment by bigs** [comment]
i'd say this is ready for review now. @Stebalien not sure who else to rope in, but this is relevant to stuff we've been thinking about

File: ds_test.go [hunk 1]
```diff
@@ -547,3 +548,155 @@ func TestDiskUsage(t *testing.T) {
 	}
 	d.Close()
 }
+
+func TestTxnRollback(t *testing.T) {
+	d, err := NewDatastore("/tmp/testing_badger_du", nil)
```
Comment (magik6k) [inline ds_test.go]: Should use `ioutil.TempDir`

File: datastore.go [hunk 2]
```diff
@@ -209,61 +300,10 @@ func (d *datastore) QueryNew(q dsq.Query) (dsq.Results, error) {
...
+	return t.txn.Commit(nil)
 }
 
-func (d *datastore) CollectGarbage() error {
-	err := d.DB.RunValueLogGC(d.gcDiscardRatio)
-	if err == badger.ErrNoRewrite {
-		err = nil
-	}
-	return err
+func (t *txn) Rollback() {
```
[the remark this answers is not in the document]
Comment (magik6k) [reply datastore.go]: Makes more sense, but still deserves a docstring explaining that it doesn't touch committed data

File: datastore.go [hunk 3]
```diff
@@ -108,42 +214,28 @@ func (d *datastore) Get(key ds.Key) (value interface{}, err error) {
...
-		return false, nil
-	}
-	if err != nil {
-		return false, err
-	}
-
-	return true, nil
-}
+func (t *txn) Has(key ds.Key) (bool, error) {
+	_, err := t.Get(key)
```
[the remark this answers is not in the document]
Comment (bigs) [reply datastore.go]: ah you know what i can open a transaction that specifically doesn't pre-load values. i will change that. good catch

File: datastore.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (schomatis) [reply datastore.go]: +1 on `Discard`, especially because the user will see it being called on every `txn` method.

**Comment by bigs** [comment]
@schomatis yeah, i do understand that. i could just duplicate the method signatures in Txn if that would help. either way, until we come to a decision on that i'll make sure the interface has good docstrings to explain.

ultimately, i think it'd be a mistake not to implement transactions in their fullest capacity just because golang makes it difficult to represent the types well. there's an awful lot of performance/efficiency to be gained!

**Comment by schomatis** [comment]
> ultimately, i think it'd be a mistake not to implement transactions in their fullest capacity just because golang makes it difficult to represent the types well. there's an awful lot of performance/efficiency to be gained!

Agreed.

**Comment by bigs** [comment]
cheers, thanks for feedback @schomatis ! change with that is up on the datastore repo

**Pull request merged**